src/blocks/battery/apc_ups.rs

```diff
@@ -0,0 +1,196 @@
+use super::{BatteryDevice, BatteryInfo, BatteryStatus, DeviceName};
+use crate::blocks::prelude::*;
+use std::io;
+use std::str::FromStr;
+use tokio::net::TcpStream;
+use tokio::time::Interval;
+
+pub(super) struct Device {
+    stream: TcpStream,
+    interval: Interval,
+}
+
+impl Device {
+    pub(super) async fn new(dev_name: DeviceName, interval: Seconds) -> Result<Self> {
+        Ok(Self {
+            stream: TcpStream::connect(dev_name.exact().unwrap_or("localhost:3551"))
+                .await
+                .error("Failed to connect to socket")?,
+            interval: interval.timer(),
+        })
+    }
+
+    async fn get_status(&self) -> Result<HashMap<String, String>> {
+        self.write(b"status").await?;
+
+        let result = self.read_response().await?;
+
+        let mut status_data: HashMap<String, String> = HashMap::new();
+        for line in result.lines() {
+            let (key, value) = line.split_once(':').unwrap();
+            status_data.insert(String::from(key.trim()), String::from(value.trim()));
+        }
+
+        Ok(status_data)
+    }
+
+    async fn write(&self, msg: &[u8]) -> Result<()> {
+        let msg_len = msg.len();
+        if msg_len >= 1 << 16 {
+            return Err(Error::new(
+                "msg is too long, it must be less than 2^16 characters long",
+            ));
+        }
+
+        // Write how many bytes of data are going to be sent (only need the last two octets)
+        // Write the actual message
+        let msgs = [&msg_len.to_be_bytes()[6..], msg];
+
+        for msg in msgs {
+            loop {
+                // Wait for the socket to be writable
+                self.stream
+                    .writable()
+                    .await
+                    .error("Socket closed unexpectedly")?;
+
+                // Try to write data, this may still fail with `WouldBlock`
+                // if the readiness event is a false positive.
+                match self.stream.try_write(msg) {
+                    Ok(_) => {
+                        break;
+                    }
+                    Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                        continue;
+                    }
+                    Err(e) => {
+                        return Err(Error::new(e.to_string()));
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn read_exact(&self, n: usize) -> Result<Vec<u8>> {
+        loop {
+            // Wait for the socket to be readable
+            self.stream
+                .readable()
+                .await
+                .error("Socket closed unexpectedly")?;
+
+            let mut buf = vec![0_u8; n];
+
+            // Try to read data, this may still fail with `WouldBlock`
+            // if the readiness event is a false positive.
+            match self.stream.try_read(&mut buf) {
+                Ok(_) => return Ok(buf),
```

Comment:
I don't think this is correct: `try_read()` is not obligated to fill the buffer completely, i.e. it can return a number less than `n`.